Add functions to dump local and global wait edges

marcocitus opened the pull request:
This currently includes the changes in #1489.

This PR adds the `dump_local_wait_edges` and `dump_global_wait_edges` function from #1447 with some cleanup and fixes around the way lock groups are handled.

anarazel commented:
@marcocitus Could you rebase, now that #1489 is merged?

marcocitus commented:
Yep, rebased.

onderkalaci commented on src/backend/distributed/transaction/lock_graph.c:
```diff
@@ -0,0 +1,772 @@
+
+	curEdge->waitingPid = waitingProc->pid;
+
+	if (IsInDistributedTransaction(waitingXactData))
```
Hmm, when a transaction is prepared, we `unset` the backend data. In that case, wouldn't we see a distributed transaction as a local one in the graph? Would that cause any problems?

marcocitus replied:
We don't really care about prepared transactions. Even though they can block other processes they cannot themselves be blocked on a lock so by definition they are not part of a deadlock. Of course, they can hold the whole world, which is why we should regularly call `recover_prepared_transactions`.

onderkalaci commented on src/backend/distributed/transaction/lock_graph.c:
```diff
@@ -0,0 +1,772 @@
+
+
+/*
+ * AddEdgesForLockWaits adds an edge to the wait graph for every lock that
```
Also, I think it is worth mentioning that the caller needs to hold the shared memory locks? (And thus do not change the function to do crazy things 😄 )

marcocitus replied:
The terminology gets a bit tricky here. Postgres source code usually refers to a "lockable object" or "lock object", and a process can hold and wait for a number of locks on the same lockable object.

onderkalaci replied:
>  AddEdgesForLockWaits adds an edge to the wait graph for every lock that waitingProc is waiting for.

I still think that this comment needs some clarification. We're only iterating through `waitingProc->waitLock-> procLocks`. So isn't it `Adds an edge to the wait graph for every process that holds the lock (that waitingProc is waiting on) in conflicting manner`

Or something with better words?

marcocitus commented on src/backend/distributed/transaction/lock_graph.c:
```diff
@@ -0,0 +1,772 @@
+	int curBackend = 0;
+	bool visited[MaxBackends];
+	PROCStack remaining;
+
```
but sourceNodeId is almost always 0 (the group ID of the coordinator).

marcocitus commented on src/backend/distributed/transaction/lock_graph.c:
```diff
@@ -0,0 +1,554 @@
+	WaitGraph *waitGraph = BuildWaitGraphForSourceNode(sourceNodeId);
+	ReturnWaitGraph(waitGraph, fcinfo);
+
+	return (Datum) 0;
```
The `return (Datum) 0;` seems to be the common pattern for SRFs in postgres, even though `PG_RETURN_VOID()` is the same. I presume it's because `PG_RETURN_VOID()` is meant for void functions.

marcocitus commented on src/backend/distributed/transaction/lock_graph.c:
```diff
@@ -0,0 +1,554 @@
+		 * that involve its own transactions.
+		 */
+		if (sourceNodeId != currentBackendData->transactionId.initiatorNodeIdentifier ||
+			!IsInDistributedTransaction(currentBackendData))
```
Moved to backend_data.c

marcocitus commented on src/backend/distributed/transaction/lock_graph.c:
```diff
@@ -0,0 +1,772 @@
+	{
+		PGPROC *waitingProc = remaining.procs[--remaining.procCount];
+
+		/*
```
Updated the comment.

marcocitus commented on src/backend/distributed/transaction/lock_graph.c:
```diff
@@ -0,0 +1,772 @@
+ */
+static void
+AddEdgesForWaitQueue(WaitGraph *waitGraph, PGPROC *waitingProc, PROCStack *remaining)
+{
```
Waiting PROCLOCKs have 0 as their holdMask (since they don't hold anything), which means that we can't check for conflict.

There's probably a way to simplify postgres though :)

anarazel replied:
> There's probably a way to simplify postgres though :)

Yea - a lot of this code is really crummy and *old*.

onderkalaci replied:
I think the only thing that annoys me in this PR is the need to have two separate iterations for doing two very similar tasks (i.e., `AddEdgesForWaitQueue()` - `AddEdgesForLockWaits ()`). 

In order to simplify things, I've a suggestion:

* `waitLock->procLocks` holds every backend that is associated with this lock. So iterate only on that.
    * For Lock Waits, do what we're already doing `procLock->holdMask & conflictMask != 0`
    * For Wait queue, do the following: `procLock->myProc->waitLock == waitLock` && `procLock->myProc->waitLockMode & conflictMask != 0`

Does the above work?

marcocitus replied:
Not really, since that would also produce wait edges for proclocks that are behind `waitingProc` in the lock queue. Only `waitQueue` contains the true order of the lock queue, which may also be modified by postgres' deadlock detection.

anarazel commented on src/backend/distributed/transaction/backend_data.c:
```diff
@@ -487,3 +487,57 @@ AssignDistributedTransactionId(void)
+UnlockDistributedTransactionSharedMemory(void)
+{
+	LWLockRelease(&backendManagementShmemData->lock);
+}
```
For me these wrappers are an abstraction that increases cost at no gains. The cost is small tho...

anarazel commented on src/backend/distributed/transaction/lock_graph.c:
```diff
@@ -0,0 +1,752 @@
+	 * Build lock-graph.  We do so by first finding all procs which we are
+	 * interested in (originating on our source system, and blocked).  Once
+	 * those are collected, do depth first search over all procs blocking
+	 * those.  To avoid redundantly dropping procs, keep track of which procs
```
"dropping procs" sounds a bit weird.

anarazel commented on src/backend/distributed/transaction/lock_graph.c:
```diff
@@ -0,0 +1,752 @@
+	int conflictMask = lockMethodTable->conflictTab[waitingProc->waitLockMode];
+
+	/* iterate through the queue of processes holding the lock */
+	SHM_QUEUE *procLocks = &(waitLock->procLocks);
```
the parens here are a bit weird.

anarazel commented on src/backend/distributed/transaction/lock_graph.c:
```diff
@@ -0,0 +1,752 @@
+		{
+			BackendData *blockingBackendData = GetBackendDataForProc(blockingProc);
+
+			if ((procLock->holdMask & conflictMask) != 0)
```
Add comment about this test - not necessarily obvious.

anarazel commented on src/backend/distributed/transaction/lock_graph.c:
```diff
@@ -0,0 +1,752 @@
+	int queueSize = waitQueue->size;
+	PGPROC *currentProc = (PGPROC *) waitQueue->links.next;
+
+	while (queueSize-- > 0 && currentProc != waitingProc)
```
Add comment saying that once we reached ourself, we've processed all locks in front of the queue?

anarazel commented on src/test/regress/specs/isolation_dump_local_wait_edges.spec:
```diff
@@ -0,0 +1,87 @@
+
+step "dist13-update"
+{
+	UPDATE local_table SET y = 3 WHERE x = 1;
```
Differently indented...

anarazel commented on src/test/regress/specs/isolation_dump_local_wait_edges.spec:
```diff
@@ -0,0 +1,87 @@
+permutation "local-begin" "dist13-begin" "local-update" "dist13-update" "detector-dump-wait-edges" "local-abort" "dist13-abort"
+
+# Distributed transaction blocked by a regular transaction blocked by a distributed transaction
+permutation "dist11-begin" "local-begin" "dist13-begin" "dist11-update" "local-update" "dist13-update" "detector-dump-wait-edges" "dist11-abort" "local-abort" "dist13-abort"
```
I was thinking for a moment that it would be good to have a bunch of test with different kind of locks (weird cases like w/ concurrently), but I can't come up with a scenario where it'd matter...

anarazel commented on src/test/regress/specs/isolation_dump_global_wait_edges.spec:
```diff
@@ -0,0 +1,85 @@
+permutation "s1-begin" "s2-begin" "s1-update" "s2-update" "detector-dump-wait-edges" "s1-abort" "s2-abort"
+
+# Distributed transaction blocked by another distributed transaction blocked by another distributed transaction
+permutation "s1-begin" "s2-begin" "s3-begin" "s1-update" "s2-update" "s3-update" "detector-dump-wait-edges" "s1-abort" "s2-abort" "s3-abort"
```
Would be good to have a local transaction in the mix - but probably too hard to make it work...

marcocitus replied:
yeh, I think it might make sense to start extending the isolation testing framework as suggested in https://github.com/citusdata/citus/issues/1499

onderkalaci replied:
> yeh, I think it might make sense to start extending the isolation testing framework as suggested in #1499

I'm also excited to write / see new tests with it.

marcocitus commented:
@onderkalaci I've cleaned up some of the redundancy between `AddEdgesForLockWaits` and `AddEdgesForWaitQueue` and added some extra comments as to why they are separate. Does this work for you?

marcocitus commented on src/backend/distributed/transaction/lock_graph.c:
```diff
@@ -0,0 +1,752 @@
+		char *nodeName = workerNode->workerName;
+		int nodePort = workerNode->workerPort;
+		MultiConnection *connection = NULL;
+		int connectionFlags = 0;
```
We're not trying to optimise for latency here, keeping connections from every node to every other node open seems costlier than opening new connections every few seconds.

onderkalaci replied:
Then, shall we use `FORCE_NEW_CONNECTION` to make this behavior consistent with every usage pattern? 

Because currently we may or may not open a new connection depending on the cached connections.

marcocitus replied:
> Because currently we may or may not open a new connection depending on the cached connections.

That seems like desirable behaviour. We don't want to explicitly persist connections in the maintenance daemon, but if you're going to call this from a Citus session and there are existing connections, then why not use them?

onderkalaci replied:
I don't have strong opinions, just to be consistent and prevent any future changes (to connection mngmnt) from breaking our desirable behavior here.

marcocitus commented on src/test/regress/expected/isolation_dump_global_wait_edges.out:
```diff
@@ -0,0 +1,85 @@
+
+waiting_transaction_numblocking_transaction_numblocking_transaction_waiting
+
+193            192            f              
```
Yeh, I thought about coming up with some CTE that maps values consistently, but it gets pretty complicated.

marcocitus commented on src/backend/distributed/transaction/lock_graph.c:
```diff
@@ -0,0 +1,752 @@
+		values[1] = Int32GetDatum(curEdge->waitingNodeId);
+		if (curEdge->waitingTransactionNum != 0)
+		{
+			values[2] = Int64GetDatum(curEdge->waitingTransactionNum);
```
This mostly determines how the user sees the output of `dump_*_wait_edges`. I think NULL is slightly prettier because it usually shows up as empty.

marcocitus commented on src/backend/distributed/transaction/lock_graph.c:
```diff
@@ -0,0 +1,769 @@
+				 errmsg(
+					 "set-valued function called in context that cannot accept a set")));
+	}
+	if (!(rsinfo->allowedModes & SFRM_Materialize))
```
Not doing this could cause crashes if the function is called from a context that doesn't support results returned using a tuple store, although I don't think such contexts exist right now. It's more a convention.

marcocitus commented on src/backend/distributed/transaction/lock_graph.c:
```diff
@@ -0,0 +1,752 @@
+static bool
+ParseBoolField(PGresult *result, int rowIndex, int colIndex)
+{
+	char *resultString = NULL;
```
Updated this to more explicitly implement boolean parsing.

onderkalaci commented on src/backend/distributed/transaction/backend_data.c:
```diff
@@ -501,3 +501,57 @@ CurrentDistributedTransactionNumber(void)
+
+
+/*
+ * LockBackendDataSharedMemory locks the shared memory segment used to keep
```
Why do we want to prevent new backends?

marcocitus replied:
Mostly to avoid using spinlocks everywhere in `dump_local_wait_edges`. Will clarify the comment a bit.
